Support async V4 payload signing

dagnir opened the pull request:
## Motivation and Context
This commit adds support for SigV4 signing of async request payloads.

In addition this commit moves the support for trailing checksums from HttpChecksumStage to the V4 signer implementation; this puts it in line with how sync chunked bodies are already handled.

This fills a gap in our signing logic where payloads that would normally be signed are not.

## Modifications

 - Remove trailing checksum path for SRA signed requests in `HttpChecksumStage`; trailing checksums taken care of in `AwsChunkedV4PayloadSigner`, same as for sync payloads
 - Support payload signing and trailers for async requests in `AwsChunkedV4PayloadSigner`
 - Internal refactor of `AwschunkedV4PayloadSigner` to reduce code duplication

## Testing
 - New unit tests
 - Enable previously disabled tests that were not passing
 - Ran integ tests

## Screenshots (if appropriate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License
<!--- The SDK is released under the Apache 2.0 license (http://aws.amazon.com/apache2.0/), so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a Contributor License Agreement (http://en.wikipedia.org/wiki/Contributor_License_Agreement) -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [x] I confirm that this pull request can be released under the Apache 2 license

dagnir committed:
Support async V4 payload signing

This commit adds support for SigV4 signing of async request payloads.

In addition this commit moves the support for trailing checksums from
HttpChecksumStage to the V4 signer implementation; this puts it in line
with how sync chunked bodies are already handled.

alextwoods reviewed (commented):
Is there a performance regression risk for customers that have been using http to avoid tls overhead?

dagnir commented on core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/util/SignerUtils.java:
```diff
@@ -221,6 +224,43 @@ public static long moveContentLength(SdkHttpRequest.Builder request, ContentStre
+     * Move `Content-Length` to `x-amz-decoded-content-length` if not already present. If `Content-Length` is not present, then
+     * the payload is read in its entirety to calculate the length.
+     */
+    public static CompletableFuture<Pair<SdkHttpRequest.Builder, Optional<Publisher<ByteBuffer>>>> moveContentLength(
```
Yeah that's fair that this could be renamed; I tried to keep the async signing path as identical to the sync signing as much as possible, and this is what what it's called on the sync path, including the consuming whole body behavior: https://github.com/aws/aws-sdk-java-v2/blob/b53211760ace3814585f467484ebca039bb99a28/core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/util/SignerUtils.java#L201

Do you have suggestions on a better name? Maybe `computeAndMoveContentLength`?

dagnir commented on core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/AwsChunkedV4PayloadSigner.java:
```diff
@@ -73,51 +78,67 @@ public static Builder builder() {
+                                                                                      .chunkSize(chunkSize)
+                                                                                      .addEmptyTrailingChunk(true);
+
+        AsyncChunkEncodedPayload checksumPayload = new AsyncChunkEncodedPayload(chunkedStreamBuilder);
```
oops, good catch. Will update the name

dagnir commented on core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/DefaultAwsV4HttpSigner.java:
```diff
@@ -230,19 +215,25 @@ private static SignedRequest doSign(SignRequest<? extends AwsCredentialsIdentity
+                                                                     V4PayloadSigner payloadSigner) {
 
         SdkHttpRequest.Builder requestBuilder = request.request().toBuilder();
+        Publisher<ByteBuffer> requestPayload = request.payload().orElse(null);
```
It would result in NPE, but yeah this shouldn't happen.

alextwoods commented on core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/chunkedencoding/ChunkedEncodedPayload.java:
```diff
@@ -0,0 +1,44 @@
+import software.amazon.awssdk.http.auth.aws.internal.signer.AwsChunkedV4PayloadSigner;
+
+/**
+ * Abstraction interface to simplify payload signing in {@link AwsChunkedV4PayloadSigner}.
```
comment nit: maybe mention it simplifies payload signing by providing an interface for both sync and async

dagnir committed:
Review comments

 - Rename to computeAndMoveContentLength
 - Rename variable to chunkedPayload
 - Fix javadocs
 - Sign payload only if non-null
